### DISCUSSION TITLE
how to mock httpfs extension in python

### DISCUSSION THREAD

**Discussion by DeoLeung** [duckdb/duckdb#12477] [General]
I'm adding duckdb to my python project and want to test reading from http

how can I mock the httpfs extension that I can capture the http request it sends and give a mocked response back?

something similar to `pytset-httpx` will be great

I checked that [python test](https://github.com/duckdb/duckdb/blob/main/tools/pythonpkg/tests/extensions/test_httpfs.py) seems to using a real object, which I want to avoid.

**Comment by Tishj** [comment]
You could probably use `fsspec`, if you register the filesystem under the same name as our own then yours should take precedence
(you can check the registered filesystems with `con.list_filesystems()`)

**Reply by DeoLeung** [reply]
yep we tried `fsspec` and substitute it during test and it works well

but still we want to test the real duckdb `httpfs` extension

currently we do it via using `http://127.0.0.1` to fetch through a local server

example local server code
```python
@app.head('/files/{file_name:str}')
@app.get('/files/{file_name:str}', response_class=FileResponse)
async def read_file(file_name: str,
                    with_hash: Annotated[bool, Query()] = False):
  if with_hash:
    # for hashed file like 8d777f385d3dfec8815d20f7496026dc_file.ext
    file_name = file_name.split('_', maxsplit=1)[1]
  path = f'tests/data/{file_name}'
  return FileResponse(path, media_type='application/octet-stream')
```
notice to implement the `HEAD` method

example pytest code
```python
@fixture
async def _fastapi_server(_free_port) -> AsyncGenerator[int, None]:
  """real uvicorn server on random free port"""
  server = UvicornTestServer(test_app, _free_port)
  await server.startup()
  yield _free_port
  server.should_exit = True
  await server.shutdown()
```

the in test just point to the `http://127.0.0.1:{_free_port}`
